homeassistant/components/binary_sensor/velbus.py

```diff
@@ -4,82 +4,62 @@
-        {
-            vol.Required('module'): cv.positive_int,
-            vol.Required('channel'): cv.positive_int,
-            vol.Required(CONF_NAME): cv.string,
-            vol.Optional('is_pushbutton'): cv.boolean
-        }
-    ])
-})
+DEPENDENCIES = ['velbus']
 
 
-def setup_platform(hass, config, add_devices, discovery_info=None):
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
     """Set up Velbus binary sensors."""
     velbus = hass.data[DOMAIN]
-
-    add_devices(VelbusBinarySensor(sensor, velbus)
-                for sensor in config[CONF_DEVICES])
+    modules = velbus.get_modules('sensor')
+    for module in modules:
+        for channel in range(1, module.number_of_channels() + 1):
+            sensor = VelbusBinarySensor(module, channel)
+            async_add_devices([sensor], update_before_add=True)
+    return True
 
 
 class VelbusBinarySensor(BinarySensorDevice):
     """Representation of a Velbus Binary Sensor."""
 
-    def __init__(self, binary_sensor, velbus):
+    def __init__(self, module, channel):
         """Initialize a Velbus light."""
-        self._velbus = velbus
-        self._name = binary_sensor[CONF_NAME]
-        self._module = binary_sensor['module']
-        self._channel = binary_sensor['channel']
-        self._is_pushbutton = 'is_pushbutton' in binary_sensor \
-                              and binary_sensor['is_pushbutton']
-        self._state = False
+        self._module = module
+        self._channel = channel
 
     @asyncio.coroutine
     def async_added_to_hass(self):
-        """Add listener for Velbus messages on bus."""
-        yield from self.hass.async_add_job(
-            self._velbus.subscribe, self._on_message)
-
-    def _on_message(self, message):
-        import velbus
-        if isinstance(message, velbus.PushButtonStatusMessage):
-            if message.address == self._module and \
-               self._channel in message.get_channels():
-                if self._is_pushbutton:
-                    if self._channel in message.closed:
-                        self._toggle()
-                    else:
-                        pass
-                else:
-                    self._toggle()
-
-    def _toggle(self):
-        if self._state is True:
-            self._state = False
-        else:
-            self._state = True
+        """Add listener for state changes."""
+        def _init_velbus():
+            """Initialize Velbus on startup."""
+            self._module.on_status_update(self._channel, self._on_update)
+        yield from self.hass.async_add_job(_init_velbus)
+
+    def _on_update(self, state):
         self.schedule_update_ha_state()
 
+    @asyncio.coroutine
+    def async_update(self):
+        """Update module status."""
+        future = self.hass.loop.create_future()
+
+        def callback():
+            """Callback when module is loaded."""
+            future.set_result(None)
```

Comment:
I think a future's result should be set from within the event loop, so schedule it with `hass.add_job`.

Reply:
ok